src/FsAutoComplete.Core/UntypedAstUtils.fs

```diff
@@ -1,668 +1,663 @@
-    =
-    addIfInside r
-    List.iter walkAttribute attrs
-    Option.iter walkTyparDecls typars
-    List.iter walkTypeConstraint constraints
-
-  and walkTypeDefnRepr =
-    function
-    | SynTypeDefnRepr.ObjectModel(_, defns, r) ->
-      addIfInside r
-      List.iter walkMember defns
-    | SynTypeDefnRepr.Simple(defn, r) ->
-      addIfInside r
-      walkTypeDefnSimple defn
-    | SynTypeDefnRepr.Exception _ -> ()
+  /// matches if the range contains the position
+  let (|IdentContainsPos|_|) pos (ident: Ident) = (|ContainsPos|_|) pos ident.idRange
 
-  and walkTypeDefnSigRepr =
-    function
-    | SynTypeDefnSigRepr.ObjectModel(_, defns, _) -> List.iter walkMemberSig defns
-    | SynTypeDefnSigRepr.Simple(defn, _) -> walkTypeDefnSimple defn
-    | SynTypeDefnSigRepr.Exception _ -> ()
-
-  and walkTypeDefn (SynTypeDefn(info, repr, members, implicitCtor, r, _)) =
-    addIfInside r
-
-    let isTypeExtensionOrAlias =
-      match repr with
-      | SynTypeDefnRepr.ObjectModel(SynTypeDefnKind.Augmentation _, _, _)
-      | SynTypeDefnRepr.ObjectModel(SynTypeDefnKind.Abbrev, _, _)
-      | SynTypeDefnRepr.Simple(SynTypeDefnSimpleRepr.TypeAbbrev _, _) -> true
-      | _ -> false
-
-    walkComponentInfo isTypeExtensionOrAlias info
-    walkTypeDefnRepr repr
-    Option.iter walkMember implicitCtor
-    List.iter walkMember members
-
-  and walkSynModuleDecl (decl: SynModuleDecl) =
-    match decl with
-    | SynModuleDecl.NamespaceFragment fragment -> walkSynModuleOrNamespace fragment
-    | SynModuleDecl.NestedModule(info, _, modules, _, r, _) ->
-      addIfInside r
-      walkComponentInfo false info
-      List.iter walkSynModuleDecl modules
-    | SynModuleDecl.Let(_, bindings, r) ->
-      addIfInside r
-      List.iter walkBinding bindings
-    | SynModuleDecl.Expr(expr, r) ->
-      addIfInside r
-      walkExpr expr
-    | SynModuleDecl.Types(types, r) ->
-      addIfInside r
-      List.iter walkTypeDefn types
-    | SynModuleDecl.Attributes(attributes = AllAttrs attrs; range = r) ->
-      addIfInside r
-      List.iter walkAttribute attrs
-    | SynModuleDecl.ModuleAbbrev(ident, longId, r) -> addIfInside r
-    | SynModuleDecl.Exception(_, r) -> addIfInside r
-    | SynModuleDecl.Open(longDotId, r) -> addIfInside r
-    | SynModuleDecl.HashDirective(_, r) -> addIfInside r
+module FoldingRange =
+  open FSharp.Compiler.Text
+  open FSharp.Compiler.Syntax
 
-  match input with
-  | ParsedInput.ImplFile input -> walkImplFileInput input
-  | _ -> ()
-  //debug "%A" idents
-  result
+  /// a walker that collects all ranges of syntax elements that contain the given position
+  [<RequireQualifiedAccess>]
+  type private RangeCollectorWalker(pos: Position) =
+    inherit SyntaxCollectorBase()
+    let ranges = ResizeArray<Range>()
+
+    let addIfInside (m: Range) =
+      if (Range.rangeContainsPos m pos) then
+        ranges.Add m
+
+    override _.WalkSynModuleOrNamespace m = addIfInside m.Range
```

Comment:
because we pulled out all of the noise of walking the tree, these individual node decision become really easy to reason about. Many AST nodes provide a helpful `Range` member so that we don't have to unpack too many of the nodes.